### ISSUE TITLE
[9.0] crash when installing l10_fr

### ISSUE THREAD

**Issue by gurneyalex** [odoo/odoo#9708]
steps to reproduce:

1. get a odoo 9 checkout (no enterprise module, but this is also reproducible with enterprise)
2. create a new empty database: `createdb test9`
3. run odoo.py -d test9 --stop-after-init -i l10n_fr
4. wait a bit. 

Expected: the installation goes smoothly

Actual:

I gent the following crash:

```
2015-11-24 09:41:10,913 458 INFO test9 openerp.modules.loading: loading l10n_generic_coa/data/configurable_account_chart.xml
2015-11-24 09:41:11,037 458 INFO test9 openerp.modules.loading: loading l10n_generic_coa/account_chart_template.yml
2015-11-24 09:41:11,043 458 INFO test9 openerp.modules.loading: loading l10n_generic_coa/../account/demo/account_bank_statement.yml
2015-11-24 09:41:11,193 458 INFO test9 openerp.modules.loading: loading l10n_generic_coa/../account/demo/account_invoice_demo.yml
2015-11-24 09:41:11,598 458 INFO test9 openerp.sql_db: Programming error: column "account_id" is of type integer but expression is of type record
LINE 1: ...e") VALUES(nextval('account_invoice_tax_id_seq'), (323, '445...
                                                             ^
HINT:  You will need to rewrite or cast the expression.
, in query INSERT INTO "account_invoice_tax" ("id", "account_id", "sequence", "invoice_id", "manual", "currency_id", "amount", "account_analytic_id", "tax_id", "name", "create_uid", "write_uid", "create_date", "write_date") VALUES(nextval('account_invoice_tax_id_seq'), %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, (now() at time zone 'UTC'), (now() at time zone 'UTC')) RETURNING id
2015-11-24 09:41:11,599 458 ERROR test9 openerp.tools.yaml_import: column "account_id" is of type integer but expression is of type record
LINE 1: ...e") VALUES(nextval('account_invoice_tax_id_seq'), (323, '445...
                                                             ^
HINT:  You will need to rewrite or cast the expression.
Traceback (most recent call last):
  File "odoo-9.0/openerp/tools/yaml_import.py", line 881, in process
    self._process_node(node)
  File "odoo-9.0/openerp/tools/yaml_import.py", line 892, in _process_node
    self.process_record(node)
  File "odoo-9.0/openerp/tools/yaml_import.py", line 334, in process_record
    module, record_dict, record_id, noupdate=self.isnoupdate(record), mode=self.mode, context=context)
  File "odoo-9.0/openerp/api.py", line 232, in wrapper
    return old_api(self, *args, **kwargs)
  File "odoo-9.0/openerp/addons/base/ir/ir_model.py", line 1107, in _update
    res_id = model_obj.create(cr, uid, values, context=context)
  File "odoo-9.0/openerp/api.py", line 232, in wrapper
    return old_api(self, *args, **kwargs)
  File "odoo-9.0/openerp/api.py", line 327, in old_api
    result = method(recs, *args, **kwargs)
  File "odoo-9.0/addons/account/models/account_invoice.py", line 317, in create
    return super(AccountInvoice, self.with_context(mail_create_nolog=True)).create(vals)
  File "odoo-9.0/openerp/api.py", line 230, in wrapper
    return new_api(self, *args, **kwargs)
  File "odoo-9.0/addons/mail/models/mail_thread.py", line 230, in create
    thread = super(MailThread, self).create(values)
  File "odoo-9.0/openerp/api.py", line 230, in wrapper
    return new_api(self, *args, **kwargs)
  File "odoo-9.0/openerp/models.py", line 4130, in create
    record = self.browse(self._create(old_vals))
  File "odoo-9.0/openerp/api.py", line 230, in wrapper
    return new_api(self, *args, **kwargs)
  File "odoo-9.0/openerp/api.py", line 463, in new_api
    result = method(self._model, cr, uid, *args, **kwargs)
  File "odoo-9.0/openerp/models.py", line 4315, in _create
    result += self._columns[field].set(cr, self, id_new, field, vals[field], user, rel_context) or []
  File "odoo-9.0/openerp/osv/fields.py", line 827, in set
    id_new = obj.create(cr, user, act[2], context=context)
  File "odoo-9.0/openerp/api.py", line 232, in wrapper
    return old_api(self, *args, **kwargs)
  File "odoo-9.0/openerp/api.py", line 327, in old_api
    result = method(recs, *args, **kwargs)
  File "odoo-9.0/openerp/models.py", line 4130, in create
    record = self.browse(self._create(old_vals))
  File "odoo-9.0/openerp/api.py", line 230, in wrapper
    return new_api(self, *args, **kwargs)
  File "odoo-9.0/openerp/api.py", line 463, in new_api
    result = method(self._model, cr, uid, *args, **kwargs)
  File "odoo-9.0/openerp/models.py", line 4268, in _create
    tuple([u[2] for u in updates if len(u) > 2])
  File "odoo-9.0/openerp/sql_db.py", line 139, in wrapper
    return f(self, *args, **kwargs)
  File "odoo-9.0/openerp/sql_db.py", line 215, in execute
    res = self._obj.execute(query, params)
ProgrammingError: column "account_id" is of type integer but expression is of type record
LINE 1: ...e") VALUES(nextval('account_invoice_tax_id_seq'), (323, '445...
                                                             ^
HINT:  You will need to rewrite or cast the expression.
```

Analysis:

The crash occurs when loading account_invoce_demo.yml, at the creation of
demo_invoice_1. The code in the yaml loader executes the on change methods, and
this creates the following value dict:

```python
{'account_id': 277,
 'amount_tax': 922.0,
 'amount_total': 5532.0,
 'amount_untaxed': 4610.0,
 'comment': False,
 'company_id': 1,
 'currency_id': 1,
 'date': False,
 'date_due': '2015-02-28',
 'date_invoice': '2015-01-01',
 'fiscal_position_id': False,
 'invoice_line_ids': [(0,
                       0,
                       {'account_analytic_id': False,
                        'account_id': 624,
                        'company_id': False,
                        'currency_id': False,
                        'discount': 0.0,
                        'invoice_line_tax_ids': [(5,), (4, 1)],
                        'name': 'Dell Inspiron Laptop without OS',
                        'price_unit': 642.0,
                        'product_id': 56,
                        'quantity': 5,
                        'uom_id': False}),
                      (0,
                       0,
                       {'account_analytic_id': False,
                        'account_id': 624,
                        'company_id': False,
                        'currency_id': False,
                        'discount': 0.0,
                        'invoice_line_tax_ids': [(5,), (4, 1)],
                        'name': 'Linutop',
                        'price_unit': 280.0,
                        'product_id': 57,
                        'quantity': 5.0,
                        'uom_id': False})],
 'journal_id': 1,
 'message_follower_ids': False,
 'message_ids': False,
 'move_id': False,
 'move_name': False,
 'name': False,
 'number': False,
 'origin': False,
 'partner_bank_id': False,
 'partner_id': 11,
 'payment_term_id': 4,
 'reconciled': True,
 'reference': False,
 'reference_type': 'none',
 'residual': False,
 'state': 'draft',
 'tax_line_ids': [(5,),
                  (0,
                   0,
                   {'account_analytic_id': False,
                    'account_id': (323,
                                   u'445711 TVA collect\xe9e (Taux Normal)'),
                    'amount': 922.0,
                    'currency_id': (1, u'EUR'),
                    'manual': False,
                    'name': u'TVA collect\xe9e (vente) 20,0%',
                    'sequence': 1,
                    'tax_id': (1, u'TVA collect\xe9e (vente) 20,0%')})],
 'type': 'out_invoice',
 'user_id': 4}
```

Note that in tax_line_ids, account_id and tax_id are not integers but
tuple. This is what is causing the crash.

**Comment by gurneyalex** [comment]
OPW-657357 opened.

**Comment by sylvain-garancher** [comment]
Running "git bisect" says : 3e0d72fa563d05cb984a9604faee9ad5939e1087 is the first bad commit
Which is not really helpful :)
At least, we know that the bug was already here on this commit.

cc @qdp-odoo

**Comment by sylvain-garancher** [comment]
With the last patch applied, git bisect gives this commit as first bad commit : https://github.com/odoo/odoo/commit/8091997e74ea6c99b11aef4a944538ce33db5c1e

**Comment by gurneyalex** [comment]
thanks to @sylvain-garancher the faulty commit seems to be https://github.com/odoo/odoo/commit/1b8c2a2078468ee040be1ebb87b9c5112c650eab 

@rco-odoo you updated the tests, the web client and fields.py but not the yaml parser, which I think is why we're having this crash.

**Comment by beledouxdenis** [comment]
My first thoughts during my analysis of this issue was simply the fact that the yaml import doesn't process correctly many2one fields into a one2many fields, that the many2one values in are left as tuple `(ID, name)`, and this lead to this issue.
https://github.com/odoo/odoo/blob/9.0/openerp/tools/yaml_import.py#L365-L397

I can be wrong, of course, I will re-analyze this issue more deeply.

**Comment by rco-odoo** [comment]
@gurneyalex @beledouxdenis you are completely right.
We changed the format of fields returned by method `onchange`, but I forgot to update `yaml_import` accordingly.

**Comment by gurneyalex** [comment]
Thanks a lot :-)

2015-12-18 16:55 GMT+01:00 Raphael Collet <notifications@github.com>:

> Closed #9708 <https://github.com/odoo/odoo/issues/9708> via #9896
> <https://github.com/odoo/odoo/pull/9896>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/odoo/odoo/issues/9708#event-495958105>.
>



-- 
Alexandre Fayolle
Chef de Projet
Tel : + 33 (0)4 58 48 20 30

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac Cedex
http://www.camptocamp.com
